### PR TITLE
RzUtil: fix rz_path_prefix in case of portable build

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Before this patch the function was using RZ_PREFIX even when IS_PORTABLE
was set, making the function not working well because RZ_PREFIX might
not exist in the system where the portable build is used (the path was
passed to realpath(), which requires an existing path).

This patch fixes the problem by searching for the BINDIR without using
realpath() on the RZ_PREFIX. Moreover, it avoid retrieving the same
prefix every time by saving it in a global variable.

**Test plan**

- Build with -Dportable=true, `eco` should list all themes
- Build with -Dportable=true --bindir=., `eco` should list all themes
- Build with -Dportable=false, `eco` should list all themes

To better test, spawn a clean container so that files are not loaded from user home dir or other system places.

@wargio Would this work for Cutter as well?

### CODE REVIEW

File: librz/util/path.c [hunk 1]
```diff
@@ -12,46 +12,102 @@
...
- * \return \p path prefixed by the Rizin install prefix or just the install prefix
- */
-RZ_API RZ_OWN char *rz_path_prefix(RZ_NULLABLE const char *path) {
 #if RZ_IS_PORTABLE
+
+static char *portable_prefix = NULL;
+static bool portable_prefix_searched = false;
+static RzThreadLock *portable_prefix_mutex = NULL;
+
+__attribute__((constructor)) static void init_portable_prefix(void) {
```
Comment (XVilka) [inline librz/util/path.c]: I think these attributes will break MSVC

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
```
FAILED: librz/util/rz_util.lib.p/thread_lock.c.obj 
"cl" "-Ilibrz\util\rz_util.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz\util\rz_util.lib.p\thread_lock.c.pdb" /Folibrz/util/rz_util.lib.p/thread_lock.c.obj "/c" ../librz/util/thread_lock.c
../librz/util/thread_lock.c(13): error C2065: 'pthread_mutexattr_t': undeclared identifier
../librz/util/thread_lock.c(13): error C2146: syntax error: missing ';' before identifier 'attr'
../librz/util/thread_lock.c(13): error C2065: 'attr': undeclared identifier
../librz/util/thread_lock.c(14): error C2065: 'attr': undeclared identifier
../librz/util/thread_lock.c(16): error C2065: 'attr': undeclared identifier
../librz/util/thread_lock.c(16): error C2065: 'PTHREAD_MUTEX_RECURSIVE': undeclared identifier
../librz/util/thread_lock.c(20): error C2065: 'attr': undeclared identifier
[114/1791] Compiling C object librz/util/rz_util.lib.p/time.c.obj
c:\projects\rizin\librz\include\rz_util/rz_num.h(166): warning C414
"cl" "-Ilibrz\util\rz_util.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz\util\rz_util.lib.p\signal.c.pdb" /Folibrz/util/rz_util.lib.p/signal.c.obj "/c" ../librz/util/signal.c
c:\projects\rizin\librz\include\rz_util/rz_num.h(166): warning C4146: unary minus operator applied to unsigned type, result still unsigned
../librz/util/signal.c(74): error C2143: syntax error: missing ')' before '*'
../librz/util/signal.c(74): error C2143: syntax error: missing '{' before '*'
../librz/util/signal.c(74): error C2143: syntax error: missing ';' before '*'
../librz/util/signal.c(74): error C2059: syntax error: ')'
../librz/util/signal.c(74): error C2054: expected '(' to follow 'oldmask'
[100/1791] Linking static target subprojects/sdb/src/libsdb.a
[101/1791] Compiling C object librz/util/rz_util.lib.p/skyline.c.obj
ninja: build stopped: subcommand failed.
Command exited with code 1
```
Windows build is broken.

**Comment by XVilka** [comment]
Can you please also fix this warning?
```
c:\projects\rizin\librz\include\rz_util/rz_num.h(166): warning C4146: unary minus operator applied to unsigned type, result still unsigned
```
It's unrelated, I know, but easy to fix and will reduce the spam from MSVC.

File: librz/include/rz_constructor.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+#define RZ_DEFINE_CONSTRUCTOR(_func) RZ_MSVC_CTOR(_func, RZ_MSVC_SYMBOL_PREFIX)
+#define RZ_DEFINE_DESTRUCTOR(_func)  RZ_MSVC_DTOR(_func, RZ_MSVC_SYMBOL_PREFIX)
+
+#define RZ_MSVC_CTOR(_func, _sym_prefix) \
+	static void _func(void); \
+	extern int (*_array##_func)(void); \
+	int _func##_wrapper(void) { \
+		_func(); \
+		g_slist_find(NULL, _array##_func); \
```
Comment (ret2libc) [inline librz/include/rz_constructor.h]: ooops

Comment (XVilka) [reply librz/include/rz_constructor.h]:
```
rz_util.lib(D__a_rizin_rizin_librz_util_path.c.obj) : error LNK2019: unresolved external symbol g_slist_find referenced in function init_portable_prefix_wrapper
```
https://github.com/rizinorg/rizin/runs/6659582878?check_suite_focus=true#step:6:2754
```
+cd test
+rz-test -L -o results.json
rz-test: error while loading shared libraries: librz_util.so.0.4: cannot open shared object file: No such file or directory
```
https://app.travis-ci.com/github/rizinorg/rizin/jobs/571833103#L3451

File: librz/include/rz_constructor.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+#define RZ_DEFINE_CONSTRUCTOR(_func) RZ_MSVC_CTOR(_func, RZ_MSVC_SYMBOL_PREFIX)
+#define RZ_DEFINE_DESTRUCTOR(_func)  RZ_MSVC_DTOR(_func, RZ_MSVC_SYMBOL_PREFIX)
+
+#define RZ_MSVC_CTOR(_func, _sym_prefix) \
+	static void _func(void); \
+	extern int (*_array##_func)(void); \
+	int _func##_wrapper(void) { \
+		_func(); \
+		char *_func##_var = rz_str_new(""); \
```
Comment (wargio) [inline librz/include/rz_constructor.h:65]: malloc 1 ?

Comment (ret2libc) [reply librz/include/rz_constructor.h:65]: I believe the purpose of calling g_slist_find initially was to use an exported function for sure. malloc could end up being optimized somehow maybe by the compiler?? Not sure.

Comment (wargio) [reply librz/include/rz_constructor.h:65]: ok makes sense..

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**